numpydoc/xref.py

```diff
@@ -0,0 +1,169 @@
+        text
+    xref_aliases : dict
+        Mapping used to resolve common abbreviations and aliases
+        to fully qualified names that can be cross-referenced.
+
+    Returns
+    -------
+    out : str
+        Text with parts that may be wrapped in a
+        ``xref_param_type`` role.
+    """
+    if param_type in xref_aliases:
+        param_type = xref_aliases[param_type]
+
+    if (QUALIFIED_NAME_RE.match(param_type) and
+            param_type not in IGNORE):
+        return ':xref_param_type:`%s`' % param_type
+
+    # Clever stuff below (except the last return)
+    # can be removed without affecting the basic functionality.
+
+    def _split_and_apply_re(s, pattern):
+        """
+        Split string using the regex pattern,
+        apply main function to the parts that do not match the pattern,
+        combine the results
+        """
+        results = []
+        tokens = pattern.split(s)
+        if len(tokens) > 1:
+            for tok in tokens:
+                if pattern.match(tok):
+                    results.append(tok)
+                else:
+                    results.append(
+                        make_xref_param_type(tok, xref_aliases))
+
+            return ''.join(results)
+        return s
+
+    def _split_and_apply_str(s, on):
+        """
+        Split string s, at the substring on,
+        apply main function to the splits,
+        combine the results
+        """
+        return on.join(
+            make_xref_param_type(s, xref_aliases)
+            for s in s.split(on))
+
+    # The cases are dealt with in an order the prevents
+    # conflict.
+    # Then the strategy is:
+    #   - Identify a pattern we are not interested in
+    #   - split off the pattern
+    #   - re-apply the function to the other parts
+    #   - join the results with the pattern
+
+    # endswith ', optional'
+    if param_type.endswith(', optional'):
+        return '%s, optional' % make_xref_param_type(
+            param_type[:-10],
+            xref_aliases)
+
+    # Any sort of bracket '[](){}'
+    has_container = any(c in CONTAINER_CHARS for c in param_type)
+    if has_container:
+        # of the form 'dict[int, float]'
+        if CONTAINER_TYPE_RE.match(param_type):
+            return _split_and_apply_re(param_type, CONTAINER_SPLIT_RE)
+        else:
+            # of the form '[int, float]'
+            for start, end in ['[]', '()', '{}']:
+                if param_type.startswith(start) and param_type.endswith(end):
+                    return '%s%s%s' % (
+                        start,
+                        make_xref_param_type(param_type[1:-1], xref_aliases),
+                        end)
+
+    # May have an unsplittable literal
+    if '``' in param_type:
+        return _split_and_apply_re(param_type, DOUBLE_QUOTE_SPLIT_RE)
+
+    # Is splittable
+    for splitter in [' or ', ', ', ' ']:
+        if splitter in param_type:
+            return _split_and_apply_str(param_type, splitter)
```

Comment:
Is there a reason we are doing this recursively for each delimiter rather than a complete tokenization step followed by handling each token?